docs/custom-plugins/custom-rules.md

```diff
@@ -0,0 +1,116 @@
+
+## Build the custom rule
+
+Each rule is a function that returns an object with methods that Redocly CLI calls to "visit" nodes while traversing the definition document. The object keys are the node types that are encountered in the document. In this simple example, the custom plugin holds a rule that fails if any `operationId` is set to "test".
+
+To keep the plugin code manageable, each rule can go in its own file. This example is in `plugins/rules/opid-not-test.js`:
+
+```js
+module.exports = OperationIdNotTest;
+
+function OperationIdNotTest() {
+  return {
+    Operation(operation, ctx) {
+      if (operation.operationId === 'test') {
+        ctx.report({
+          message: `operationId must be not "test"`,
+          location: ctx.location.child('operationId'),
+        });
+      }
+    },
+  };
+}
+
+```
+
+The `ctx` object here holds all the context, which can be used to give more situation-aware functionality to the rules you build. This is one of the main use cases for custom rules. The `report()` method is used to give information to return to the user if the node being visited doesn't comply with the rule. You can read the [context](#the-context-object) and [location](#location-object) sections for more information.
+
+Adding this as part of a plugin requires you to add it to the `rules` part of the plugin object, under the relevant document type. The example rule here is intended to be used with OpenAPI, so the plugin code in `plugins/my-rules.js` is as follows:
+
+```
+const OpIdNotTest = require('./rules/opid-not-test.js');
+
+module.exports = {
+  id: 'my-rules',
+  rules: {
+    oas3: {
+      'opid-not-test': OperationIdNotTest,
+    }
+  },
+};
+```
+
+To use the example rule, add the following to your `redocly.yaml` configuration file:
+
+```yaml
+plugins:
+  - 'plugins/my-rules.js'
+
+rules:
+  my-rules/opid-not-test: warn
+```
+
+Validate your OpenAPI document with `redocly lint openapi.yaml`. With this rule enabled as shown, any `operationId` fields that are "test" cause the validation step to emit a warning. You can use the example here as a basis for building your own rules, and many rules can be included in a single plugin.
+
+## Object references
+
+### The context object
+
+The context object contains additional functionality that is helpful for rules to do their jobs. As the name implies, the context object contains information that is relevant to the context of the rule. The context object has the following properties:
+
+- `location` - current location in the source document. See [Location Object](#location-object)
+- `parentLocations` - mapping of parent node to its location (only for nested visitors)
+- `type` - information about current type from type tree
+- `parent` - parent object or array
+- `key` - key in parent object or array
+- `oasVersion` specific OAS minor version of current document (can be `oas2`, `oas3` or `oas3_1`).
+
+The context object also offers some additional functionality that 
```

Comment:
This is odd - perhaps this was added as a mistake?

Reply:
Apparently my brain stalled mid-sentence? Thanks for catching this